### PR TITLE
Feature/improve fixed

### PR DESCRIPTION
**Pull request by x-zho14**

### CODE REVIEW

File: common/fixed.go [hunk 1]
```diff
@@ -63,86 +88,135 @@ func (f *Fixed) shrinkDecimal() *Fixed {
...
 // Equals check equal
 func (f *Fixed) Equals(other *Fixed) bool {
-	fpnNew, otherNew := UnifyDecimal(f, other)
+	fpnNew, otherNew, err := UnifyDecimal(f, other)
+	f.err = err
 	return fpnNew.Value == otherNew.Value
 }
 
 // Add ...
 func (f *Fixed) Add(other *Fixed) *Fixed {
```
Comment (liangjn16) [inline common/fixed.go]: return *Fixed, err

File: common/fixed.go [hunk 2]
```diff
@@ -63,86 +88,135 @@ func (f *Fixed) shrinkDecimal() *Fixed {
...
+			if fpn.Value < 0 {
+				return nil, errOverflow
+			}
+			if decimalStart {
+				fpn.Decimal++
+				if fpn.Decimal >= decimal {
+					break
+				}
+			}
+		} else if amount[i] == '.' {
```
Comment (liangjn16) [inline common/fixed.go]: if amount has tow '.'

File: vm/database/token_handler.go [hunk 3]
```diff
@@ -54,8 +54,8 @@ func (m *TokenHandler) SetTokenBalance(tokenName, acc string, amount int64) {
 
 // SetTokenBalanceFixed set token balance of acc, used for test
 func (m *TokenHandler) SetTokenBalanceFixed(tokenName, acc string, amountStr string) {
-	amountNumber, ok := common.NewFixed(amountStr, m.Decimal(tokenName))
-	if !ok {
+	amountNumber, err := common.NewFixed(amountStr, m.Decimal(tokenName))
+	if err != nil {
 		panic(errors.New("construct Fixed number failed. str = " + amountStr + ", decimal = " + string(m.Decimal(tokenName))))
```
Comment (liangjn16) [inline vm/database/token_handler.go]: panic?

File: vm/host/teller.go [hunk 4]
```diff
@@ -44,7 +44,6 @@ func (h *Teller) TransferRawNew(from, to string, amount int64) error {
 	return ErrBalanceNotEnough
 }
 
-// GrantCoin issue coin. todo deprecated
```
Comment (liangjn16) [inline vm/host/teller.go]: golint

File: vm/monitor.go [hunk 5]
```diff
@@ -97,8 +97,8 @@ func (m *Monitor) Call(h *host.Host, contractName, api string, jarg string) (rtn
 		cost.AddAssign(cost0)
 		for _, limit := range amountLimit {
 			decimal := h.DB().Decimal(limit.Token)
-			fixedAmount, ok := common.NewFixed(limit.Val, decimal)
-			if ok {
+			fixedAmount, err := common.NewFixed(limit.Val, decimal)
+			if err == nil {
```
[the remark this answers is not in the document]
Comment (wangyu0air) [reply vm/monitor.go]: will check amountLimit when deploy contract

File: vm/database/token_handler.go (hunk 3 above)
Comment (wangyu0air) [reply vm/database/token_handler.go]: this function is used for test

**Comment by lispc** [comment]
ok

**Pull request merged**